### PR TITLE
Date picker review2

### PR DESCRIPTION
**Pull request by mark-tate**

### CODE REVIEW

File: site/src/examples/date-picker/SingleWithCustomPanel.tsx [hunk 1]
```diff
@@ -14,13 +14,13 @@ import {
   DatePickerSingleInput,
   type SingleDateSelection,
   formatDate,
-  getCurrentLocale,
+  getCurrentLocale, SingleDatePickerError,
 } from "@salt-ds/lab";
 import { CustomDatePickerPanel } from "@salt-ds/lab/stories/date-picker/CustomDatePickerPanel";
```
Comment (origami-z) [inline site/src/examples/date-picker/SingleWithCustomPanel.tsx:19]: SO 92803, these aren't exports, maybe do something like https://github.com/jpmorganchase/salt-ds/blob/0c3721a43095c759f15398c986727179b689275d/site/src/examples/ag-grid-theme/CheckboxSelection.tsx#L2-L3

**Comment by origami-z** [comment]
> 🫰✨ **Thanks @origami-z! Your snapshots have been published to npm.**
> 
> Test the snapshots by updating your `package.json` with the newly published versions:
> 
> ```shell
> yarn add @salt-ds/ag-grid-theme@0.0.0-snapshot-20241103162723
> ```
> 
> ```shell
> yarn add @salt-ds/core@0.0.0-snapshot-20241103162723
> ```
> 
> ```shell
> yarn add @salt-ds/countries@0.0.0-snapshot-20241103162723
> ```
> 
> ```shell
> yarn add @salt-ds/data-grid@0.0.0-snapshot-20241103162723
> ```
> 
> ```shell
> yarn add @salt-ds/lab@0.0.0-snapshot-20241103162723
> ```
> 
> ```shell
> yarn add @salt-ds/theme@0.0.0-snapshot-20241103162723
> ```

Slowly adding tests on  https://stackblitz.com/edit/salt-template-xqddfx?file=App.tsx

File: packages/lab/stories/date-picker/date-picker.stories.tsx [hunk 2]
```diff
@@ -784,359 +1265,703 @@ export const RangeWithConfirmation: StoryFn<DatePickerRangeProps> = ({
...
+      </DatePicker>
+      {!open ? <FormHelperText>{helperText}</FormHelperText> : null}
+    </FormField>
+  );
+};
 
-  function renderDayContents(day: DateValue) {
-    const formatter = new DateFormatter("en-US", { day: "numeric" });
-    return <>{formatter.format(day.toDate(getLocalTimeZone()))}</>;
+export const SingleWithLocaleZhCN: StoryFn<
```
Comment (origami-z) [inline packages/lab/stories/date-picker/date-picker.stories.tsx:1716]: Haven't look closely, i think the first day of week is not working

File: packages/lab/src/calendar/internal/CalendarDay.tsx [hunk 3]
```diff
@@ -52,7 +55,7 @@ export const CalendarDay = forwardRef<HTMLButtonElement, CalendarDayProps>(
 
     const dayRef = useRef<HTMLButtonElement>(null);
     const buttonRef = useForkRef(ref, dayRef);
-    const { status, dayProps, unselectableReason, highlightedReason, locale } =
+    const { status, dayProps, unselectableReason, highlightedReason } =
```
Comment (origami-z) [inline packages/lab/src/calendar/internal/CalendarDay.tsx:58]: Unselectable and disabled styles are potentially wrong (shown as single horizontal stroke). One should be a cross and one should be no strikethrough?

File: packages/lab/src/date-picker/DatePickerRangePanel.tsx [hunk 4]
```diff
@@ -352,12 +354,12 @@ export const DatePickerRangePanel = forwardRef<
           <Calendar selectionVariant={"range"} {...StartCalendarProps}>
```
Comment (origami-z) [inline packages/lab/src/date-picker/DatePickerRangePanel.tsx:354]: What's the best way to do single calendar for a range picker?

File: packages/lab/src/date-picker/DatePickerSingleInput.tsx [hunk 5]
```diff
@@ -28,113 +28,167 @@ const withBaseName = makePrefixer("saltDatePickerSingleInput");
...
+    return (
+      <DateInputSingle
+        value={value || ""}
+        className={clsx(withBaseName(), className)}
+        date={selectedDate || null}
+        readOnly={readOnly}
+        ref={ref}
+        onDateChange={handleDateChange}
+        onDateValueChange={handleDateValueChange}
+        endAdornment={
```
Comment (origami-z) [inline packages/lab/src/date-picker/DatePickerSingleInput.tsx:178]: Should calendar icon button be part of Date input (more like combobox)? e.g., if I pass in a clear button as `endAdornment`, then calendar icon is gone

File: packages/lab/src/calendar/CalendarWeekHeader.tsx [hunk 6]
```diff
@@ -1,28 +1,32 @@
...
-  { className, locale = getCurrentLocale(), ...rest },
-  ref,
-) {
-  const weekdaysShort = daysForLocale("narrow", locale);
-  const weekdaysLong = daysForLocale("long", locale);
+export const CalendarWeekHeader = forwardRef(function CalendarWeekHeader<
+  TDate extends DateFrameworkType,
+>(props: CalendarWeekHeaderProps, ref: React.Ref<HTMLDivElement>) {
+  const { className, ...rest } = props;
+  const { dateAdapter } = useLocalization<TDate>();
```
Comment (origami-z) [inline packages/lab/src/calendar/CalendarWeekHeader.tsx:23]: I don't think first day of week is handled by adaptor. Also week shorthand (e.g. S/M/T), is being translated by locale.

Comment (origami-z) [reply packages/lab/src/calendar/CalendarWeekHeader.tsx:23]: Confirmed working in `0.0.0-snapshot-20241106085302`

File: packages/lab/stories/date-picker/date-picker.stories.tsx (hunk 2 above)
Comment (origami-z) [reply packages/lab/stories/date-picker/date-picker.stories.tsx:1717]: Confirmed working in `0.0.0-snapshot-20241106085302`

File: packages/lab/src/date-input/DateInputSingle.tsx [hunk 7]
```diff
@@ -224,27 +200,21 @@ export const DateInputSingle = forwardRef<HTMLDivElement, DateInputSingleProps>(
...
     // Update date string value when selected date changes
     useEffect(() => {
-      const formattedDate = format(date || null);
-      if (formattedDate) {
-        setDateValue(formattedDate);
-        onDateValueChange?.(formattedDate, true);
+      if (date && dateAdapter.isValid(date)) {
+        const formattedValue = dateAdapter.format(date, format, locale);
+        const hasValueChanged = formattedValue !== dateValue;
+        if (hasValueChanged) {
```
Comment (origami-z) [inline packages/lab/src/date-input/DateInputSingle.tsx:212]: This prevents controlling `selectedDate` to `null` to reset input to empty. As Kris suggested, maybe consider using "invalid" date in `onSelectionChange`'s `date`, so there will be 4 distinct choices
- `undefined` (controlled vs uncontrolled)
- `null`, controlled with nothing
- invalid date, controlled with date but invalid
- valid date, controlled with date and also valid

File: packages/lab/src/calendar/CalendarNavigation.tsx [hunk 8]
```diff
@@ -171,55 +175,62 @@ function useCalendarNavigation() {
...
+    [visibleMonth],
   );
 
-  const years: DateValue[] = useMemo(
+  const years: TDate[] = useMemo(
     () =>
-      generateYearsBetweenRange(
-        Math.min(minDate ? minDate.year : CALENDAR_MIN_YEAR, visibleMonth.year),
-        Math.max(maxDate ? maxDate.year : CALENDAR_MAX_YEAR, visibleMonth.year),
+      generateYearsBetweenRange<TDate>(
```
Comment (origami-z) [inline packages/lab/src/calendar/CalendarNavigation.tsx:193]: Confirming this fixes year range not respecting min/max years (SO 93573)

File: packages/lab/src/date-picker/DatePickerOverlayProvider.tsx [hunk 9]
```diff
@@ -107,12 +123,26 @@ export const DatePickerOverlayProvider: React.FC<
...
+        reason === "escape-key" ||
+        (reason === "outside-press" && onDismissCallback.current)
+      ) {
+        onDismissCallback?.current?.();
+      }
+    },
+    [onOpen],
+  );
 
   const floatingUIResult = useFloatingUI({
```
Comment (origami-z) [inline packages/lab/src/date-picker/DatePickerOverlayProvider.tsx:147]: Placement is hard coded here, can we expose things like `placement` from `Tooltip`? SO 69481

**Comment by origami-z** [comment]
> 🫰✨ **Thanks @origami-z! Your snapshots have been published to npm.**
> 
> Test the snapshots by updating your `package.json` with the newly published versions:
> 
> ```shell
> yarn add @salt-ds/ag-grid-theme@0.0.0-snapshot-20241106085302
> ```
> 
> ```shell
> yarn add @salt-ds/core@0.0.0-snapshot-20241106085302
> ```
> 
> ```shell
> yarn add @salt-ds/countries@0.0.0-snapshot-20241106085302
> ```
> 
> ```shell
> yarn add @salt-ds/data-grid@0.0.0-snapshot-20241106085302
> ```
> 
> ```shell
> yarn add @salt-ds/lab@0.0.0-snapshot-20241106085302
> ```
> 
> ```shell
> yarn add @salt-ds/theme@0.0.0-snapshot-20241106085302
> ```

Tested in https://stackblitz.com/edit/salt-template-xqddfx?file=App.tsx

File: packages/lab/src/date-input/DateInputSingle.tsx [hunk 10]
```diff
@@ -181,21 +157,19 @@ export const DateInputSingle = forwardRef<HTMLDivElement, DateInputSingleProps>(
       onClick,
       emptyReadOnlyMarker = "—",
       endAdornment,
-      format: formatProp,
+      format = "DD MMM YYYY",
       inputProps = {},
       inputRef: inputRefProp = null,
-      parse = parseCalendarDate,
+      locale,
       placeholder = "dd mmm yyyy",
```
[the remark this answers is not in the document]
Comment (mark-tate) [reply packages/lab/src/date-input/DateInputSingle.tsx:164]: parse no longer exists as a prop and is owned by adapter

[the remark this answers is not in the document]
Comment (origami-z) [reply packages/lab/src/date-input/DateInputSingle.tsx:164]: The comment here is for `placeholder`. On `parse` on adapter, need to consider the user case of different date pickers in a single form, does that mean multiple adapter..?

File: site/src/examples/date-picker/SingleWithCustomPanel.tsx (hunk 1 above)
Comment (mark-tate) [reply site/src/examples/date-picker/SingleWithCustomPanel.tsx:16]: will pickup in site docs updates

File: packages/lab/src/date-input/DateInputSingle.tsx (hunk 7 above)
Comment (mark-tate) [reply packages/lab/src/date-input/DateInputSingle.tsx:214]: done

File: packages/lab/src/localization-provider/LocalizationProvider.tsx [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+ * @returns The localization provider value, which includes date manipulation and formatting utilities.
+ *
+ * @throws Will throw an error if the hook is used outside of a `LocalizationProviderContext.Provider`.
+ */
+export const useLocalization = <
+  TDate extends DateFrameworkType,
+>(): LocalizationProviderValue<TDate> => {
+  const localization = useContext(LocalizationProviderContext);
+  if (!localization) {
+    throw new Error(
```
Comment (origami-z) [inline packages/lab/src/localization-provider/LocalizationProvider.tsx:131]: I think we need to make this like `useIcon` from `SemanticIconProvider` where one of 4 adapters is the default.

**Review by origami-z — COMMENTED** [review]
We will need some blurb around `LocalizationProvider`, we could get a lot of support questions with only the date picker doc

File: packages/lab/src/calendar/CalendarNavigation.tsx (hunk 8 above)
Comment (mark-tate) [reply packages/lab/src/calendar/CalendarNavigation.tsx:191]: yes, min/max date affects the values populated in years.
There is a edge-case where your visible year is currently 2025, and you set the min/max to 2023, the years will currently show 2025, as changing the min/max does not affect the current visible month. However, once you move back to a year within the range, the years dropdown removes 2025 as a possible year. Will follow up on this, as this PR, is know in a known state. @ivan-calderon have a play with this edge case

File: packages/lab/src/calendar/internal/CalendarDay.tsx (hunk 3 above)
Comment (mark-tate) [reply packages/lab/src/calendar/internal/CalendarDay.tsx:59]: will follow up on this item in a  new PR

File: packages/lab/src/date-input/DateInputSingle.tsx [hunk 12]
```diff
@@ -110,13 +90,17 @@ export interface DateInputSingleProps<T = SingleDateSelection>
...
   /**
-   * Function to format the input value.
+   * Format string for date.
    */
-  format?: (date: DateValue | null) => string;
+  format?: string;
   /**
    * Reference for the input.
    */
   inputRef?: RefObject<HTMLInputElement>;
```
[the remark this answers is not in the document]
Comment (mark-tate) [reply packages/lab/src/date-input/DateInputSingle.tsx:99]: updated

File: packages/lab/src/date-picker/DatePickerOverlayProvider.tsx (hunk 9 above)
Comment (mark-tate) [reply packages/lab/src/date-picker/DatePickerOverlayProvider.tsx:147]: good shout, follow-up PR though... Ideal target state would be an equivalent, common component 
For now, will make this component support placement

@joshwooding previously discussed

File: packages/lab/src/localization-provider/LocalizationProvider.tsx (hunk 11 above)
Comment (mark-tate) [reply packages/lab/src/localization-provider/LocalizationProvider.tsx:131]: Needs a conversation around SaltProvider and a follow up PR

File: packages/lab/src/date-picker/DatePickerSingleInput.tsx (hunk 5 above)
Comment (origami-z) [reply packages/lab/src/date-picker/DatePickerSingleInput.tsx:183]: not asking to couple them, could do similar to ComboBox to PillInput, it automatically adds the button for `endAdornment`
https://github.com/jpmorganchase/salt-ds/blob/7d3251821e4b93cdf1eab55a99ec9a5213dfe296/packages/core/src/combo-box/ComboBox.tsx#L415-L419

File: packages/lab/src/localization-provider/LocalizationProvider.tsx (hunk 11 above)
Comment (origami-z) [reply packages/lab/src/localization-provider/LocalizationProvider.tsx:131]: If we don't do this, we need to make sure there is clear message on top of each component doc, that localization provider is needed to make them work.

File: packages/date-adapters/src/types.ts [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+  date: TDate;
+} & DateDetail;
+
+/**
+ * Interface for a date adapter, providing methods for date manipulation and formatting.
+ *
+ * @template TDate - The type of the date object used by the adapter.
+ * @template TLocale - The type of the locale, defaulting to any.
+ */
+export interface SaltDateAdapter<
```
Comment (origami-z) [inline packages/date-adapters/src/types.ts:110]: Not 100% sure why certain function can accept optional `locale` as final arg, while the others doesn't.

File: packages/date-adapters/src/__tests__/moment.spec.ts [hunk 14]
```diff
@@ -0,0 +1,179 @@
+import moment from "moment";
+import { describe, expect, it } from "vitest";
+import "moment-timezone";
+import { AdapterMoment, DateDetailErrorEnum } from "../index";
+
+describe("GIVEN a AdapterMoment", () => {
+  const adapter = new AdapterMoment({ locale: "en" });
+
+  it("should create a Moment date in the system timezone", () => {
```
Comment (origami-z) [inline packages/date-adapters/src/__tests__/moment.spec.ts:9]: system timezone in test may cause issue when running locally...?

**Review by origami-z — APPROVED** [review]

File: packages/lab/src/date-picker/DatePickerRangePanel.tsx (hunk 4 above)
Comment (mark-tate) [reply packages/lab/src/date-picker/DatePickerRangePanel.tsx:363]: it's possible, should be as simple as adding a prop to the panel that specifies the number of calendars in the view
Then you need to do something about navigation, again should be simple, as long as rendered calendars are all based of the visible month. Will revisit in another PR/example

File: packages/lab/stories/date-picker/date-picker.stories.tsx [hunk 15]
```diff
@@ -784,359 +1265,703 @@ export const RangeWithConfirmation: StoryFn<DatePickerRangeProps> = ({
...
+  );
+
+  const isDayUnselectable = useCallback(
+    (date: DateFrameworkType): string | false => {
+      return isWeekend(date) ? "weekends are un-selectable" : false;
+    },
+    [],
+  );
+
+  const validateAndCustomizeError = useCallback(
```
[the remark this answers is not in the document]
Comment (mark-tate) [reply packages/lab/stories/date-picker/date-picker.stories.tsx:1678]: will revisit in smaller pr

**Review by Fercas123 — APPROVED** [review]

**Pull request merged**